File: C++/21.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
/* Version1 One by one
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* prehead = new ListNode(0);
        ListNode* res = prehead;
        while(l1&&l2){
            if(l1->val<l2->val) {
                res->next = l1;
                l1 = l1->next;
            }
            else {
                res->next = l2;
                l2 = l2->next;
            }
            res = res->next;
        }
        if(l1) res->next = l1;
        else res->next = l2;
        return prehead->next;
    }
*/
// Version 2 Switch 
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* prehead = new ListNode(0);
        prehead->next = l1;
        ListNode* curr = prehead;
        while(l2){
            while(curr->next && curr->next->val <= l2->val) curr = curr->next;
            l1 = curr->next;
            curr->next = l2;
            l2 = l1;
        }
        return prehead->next;
    }
};

File: Python/21.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} l1
    # @param {ListNode} l2
    # @return {ListNode}
    def mergeTwoLists(self, l1, l2):
        dummy = ListNode(0)
        current = dummy
        while l1 and l2:
            if l1.val < l2.val:
                 current.next = l1
                 l1 = l1.next
            else:
                current.next = l2
                l2 = l2.next
            current = current.next
        if l1:
            current.next = l1
        if l2:
            current.next =l2
        return dummy.next